aclhound/targets/deploy_junos.py

```diff
@@ -0,0 +1,113 @@
+# this list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+# this list of conditions and the following disclaimer in the documentation
+# and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""
+Upload ACLs to Cisco IOS devices
+
+The following process is followed to ensure zero impact
+
+    upload new ACL via SFTP
+    load new ACL via SSH
+
+"""
+
+from Exscript.util.interact import read_login
+from Exscript.protocols import SSH2, Telnet, Account
+
+import netrc
+import paramiko
+
+from pprint import pprint
+from aclhound import textfsm
+from StringIO import StringIO
+
+from tempfile import NamedTemporaryFile
+
+
+def deploy(hostname=None, acls=None, transport='ssh', save_config=False,
+           timeout=60):
+    """
+    Deploy code to a JunOS device
+    """
+    try:
+        username, acc, password = \
+            netrc.netrc().authenticators(hostname)
+        account = Account(name=username, password=password, key=None)
+    except Exception, e:
+        print e
+        print("ERROR: could not find device in ~/.netrc file")
+        print("HINT: either update .netrc or enter username + pass now.")
+        try:
+            account = read_login()
+        except EOFError:
+            print("ERROR: could not find proper username + pass")
+            print("HINT: set username & pass in ~/.netrc for device %s"
+                  % hostname)
+            import sys
+            sys.exit(2)
+
+    def s(conn, line):
+        print("   %s" % line)
+        conn.execute(line)
+
+    # main flow of the program starts here
+
+    for policy in acls:
+        # deal with ipv4 only for now
+        if acls[policy]['afi'] == 4:
+            f = NamedTemporaryFile(delete=True)
+            f.write(acls[policy]['policy'])
+            tr = paramiko.Transport((hostname, 22))
+            tr.connect(username = username, password = password)
+
+            sftp = paramiko.SFTPClient.from_transport(tr)
+            sftp.put(f.name, "./{}".format(policy))
+            sftp.close()
+
+            tr.close()
+            f.close()
+
+    if save_config == True:
```

Comment:
This conditional might not be useful on JunOS - afaik the moment you `commit` on JunOS - it also becomes the startup configuration. I don't think there is a trivial or publicly well-understood way to have a difference between the 'runnning' config and the 'startup' config on JunOS. I'd assume on JunOS that save_config is always true